language-support/ts/codegen/tests/ts/daml-json-types/src/index.ts

```diff
@@ -207,18 +207,31 @@ export const ContractId = <T>(_t: Serializable<T>): Serializable<ContractId<T>>
 });
 
 /**
- * The counterpart of DAML's `Optional T` type. Nested optionals are not yet
- * supported.
- */
-export type Optional<T> = T | null;
+ * The counterpart of DAML's `Optional T` type.
+ */
+export type Optional<T> =
+  | null
+  | Optional.Inner<T>
+// eslint-disable-next-line @typescript-eslint/no-namespace
+namespace Optional {
+  export type Inner<T> = null extends T ? ([] | [Exclude<T, null>]) : T;
+}
 
 /**
- * Companion object of the `Optional` type.
+ * Companion function of the `Optional` type.
  */
 export const Optional = <T>(t: Serializable<T>): Serializable<Optional<T>> => ({
-  decoder: () => jtv.oneOf(jtv.constant(null), t.decoder()),
-  isOptional: false,
+  decoder : () => jtv.oneOf<Optional<T>>(jtv.constant(null), Optional.Inner(t)),
+  isOptional: true,
 });
+Optional.Inner = <T>(t: Serializable<T>): jtv.Decoder<Optional.Inner<T>> =>
+  ! t.isOptional
+  ? t.decoder() as jtv.Decoder<Optional.Inner<T>>
+  : jtv.oneOf(
+      jtv.constant([]) as jtv.Decoder<[] | [Exclude<T, null>]>,
+      jtv.tuple([t.decoder()]) as jtv.Decoder<[] | [Exclude<T, null>]>
```

Comment:
This is unfortunately too permissive. I'll merge it nevertheless and fix it and add a few tests in a follow-up PR.